Encryption.md

```diff
@@ -19,216 +19,544 @@
+calculate the authentication tag. Actually, GCM is a combination of CTR cipher and an 
+authentication layer  called GMAC. For applications running without AES acceleration 
+and willing to compromise on content verification, CTR cipher can provide a boost in 
+Parquet write/read throughput. The second Parquet algorithm encrypts the data content 
+(pages) with CTR, and the metadata (Thrift structures) with GCM. This allows to encrypt/decrypt 
+the bulk of the data faster, while still verifying the metadata integrity and making 
+sure the file has not been replaced with a wrong version. However, tampering with the 
+page data might go unnoticed.
+
+The initialization vectors (IVs) in GCM and CTR ciphers must be unique for each encrypted 
+stream. The CTR IVs are comprised of two parts: a unique nonce (“number used once”) and 
+an initial counter field; both must be passed to the CTR cipher upon stream initialization. 
+The GCM mode creates a CTR cipher internally, and passes an IV comprised of a nonce (passed to 
+the GCM cipher upon stream initialization) and an initial counter field constructed internally 
+by the GCM cipher according to the GCM specification (NIST SP 800-38D).
+Parquet encryption uses the RBG-based (random bit generator) nonce construction as defined in 
+the NIST SP 800-38D document, section 8.2.2. Notice: NIST SP 800-38D uses a term “IV” for what 
+is called “nonce” in the Parquet encryption design.
+
+An input to AES CTR encryption is an encryption key, a 16-byte IV and a plaintext. The output
+is a ciphertext with the length equal to that of plaintext (for non-padding encryption).
+
+An input to AES GCM encryption is an encryption key, a 12-byte nonce and a plaintext. The output
+is a ciphertext with the length equal to that of plaintext (for non-padding encryption), and 
+a 16-byte authentication tag used to verify the ciphertext integrity.
+
+
+### 4.1 AES_GCM_V1 encryption algorithm
+This Parquet algorithm encrypts all modules by the GCM cipher, without padding. The AES GCM cipher
+must be implemented by a cryptographic provider according to the NIST SP 800-38D specification. 
+
+Unique nonces are generated for each ciphertext. The nonce length is 12 bytes (96 bits). The NIST 
+specification requires a “random field” to fill all 12 bytes and an arbitrary “free field” 
+to be empty for this nonce length (section 8.2.2). Parquet uses the option 1 for RGB implementation - 
+"The random field shall .. consist of an output string of r(i) bits from an approved RBG with a 
+sufficient security strength". That is, a fresh random byte array must be generated for each nonce - 
+as opposed to the option 2 in the same spec - "or the result of applying the r(i)–bit incrementing 
+function to the random field of the preceding <nonce> for the given key"
+
+
+### 4.2 AES_GCM_CTR_V1 encryption algorithm
+In this Parquet algorithm, all Thrift modules are encrypted with the GCM cipher, as described above. 
```

Comment:
This should note "but page modules are encrypted with the CTR cipher, without padding". A separate paragraph is fine to define how CTR is used, but the difference between AES_GCM_V1 and AES_GCM_CTR_V1 should not be split across two paragraphs.